### PR TITLE
USHIFT-285: MicroShift: allow CLI creation without OpenShift APIs

### PR DESCRIPTION
**Pull request by pacevedom**
Swaps project API with k8s namespaces in test/extended/util package. Automatic detection of project API to create a CLI object according to whats available in the cluster.

### CODE REVIEW

File: test/extended/util/client.go [hunk 1]
```diff
@@ -337,6 +345,40 @@ func (c *CLI) SetupProject() string {
...
+			Name: newNamespace,
+		},
+	}
+	framework.Logf("Creating namespace %q", newNamespace)
+	_, err := c.AdminKubeClient().CoreV1().Namespaces().Create(context.Background(), nsObject, metav1.CreateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	c.kubeFramework.AddNamespacesToDelete(nsObject)
+
+	framework.Logf("Waiting on permissions in project %q ...", newNamespace)
```
Comment (dhellmann) [inline test/extended/util/client.go:364]:
```suggestion
	framework.Logf("Waiting on permissions in namespace %q ...", newNamespace)
```

File: test/extended/util/client.go [hunk 2]
```diff
@@ -770,52 +812,60 @@ func (c *CLI) CreateUser(prefix string) *userv1.User {
...
+			UserUID:     string(user.UID),
+			Scopes:      []string{"user:full"},
+			RedirectURI: "https://localhost:8443/oauth/token/implicit",
+		}, metav1.CreateOptions{})
+		if err != nil {
+			FatalErr(err)
+		}
+		c.AddResourceToDelete(oauthv1.GroupVersion.WithResource("oauthaccesstokens"), token)
+		userClientConfig.BearerToken = privToken
+	} else {
```
Comment (dhellmann) [inline test/extended/util/client.go:866]: The change in this function would be smaller if the `if` test on line 821 was reversed and then the userClientConfig being created here was returned early. That would also make the function easier to follow, because it would be clear that most of the steps apply when the users API is present.

**Review by dhellmann — COMMENTED** [review]
A couple of small suggestions, but this looks good generally.

File: test/extended/util/client.go (hunk 1 above)
Comment (pacevedom) [reply test/extended/util/client.go]: Done

File: test/extended/util/client.go (hunk 2 above)
Comment (pacevedom) [reply test/extended/util/client.go]: You are right, looks better now. Thanks!

**Comment by ggiguash** [comment]
/retitle USHIFT-285: MicroShift: allow CLI creation without OpenShift APIs

File: test/extended/util/client.go [hunk 3]
```diff
@@ -337,6 +345,40 @@ func (c *CLI) SetupProject() string {
 	return newNamespace
 }
 
+func (c *CLI) setupNamespace() string {
+	requiresTestStart()
+	newNamespace := names.SimpleNameGenerator.GenerateName(fmt.Sprintf("e2e-test-%s-", c.kubeFramework.BaseName))
+	c.SetNamespace(newNamespace).ChangeUser(fmt.Sprintf("%s-user", newNamespace))
```
Comment (deads2k) [inline test/extended/util/client.go:351]: ChangeUser requires the users.user API.

Comment (deads2k) [inline test/extended/util/client.go:351]: 1. you should try to execute this and see it work without the user API
2. you could create a new ChangeUserWithClientCert that uses a client CSR to change the identity and persist in a kubeconfig file.

File: test/extended/util/client.go [hunk 4]
```diff
@@ -337,6 +345,40 @@ func (c *CLI) SetupProject() string {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name: newNamespace,
+		},
+	}
+	framework.Logf("Creating namespace %q", newNamespace)
+	_, err := c.AdminKubeClient().CoreV1().Namespaces().Create(context.Background(), nsObject, metav1.CreateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	c.kubeFramework.AddNamespacesToDelete(nsObject)
+
```
Comment (deads2k) [inline test/extended/util/client.go:363]: you need to do the work of actually creating the things in teh standard openshfit project template like the proper annotations and rolebindings.

File: test/extended/util/client.go [hunk 5]
```diff
@@ -337,6 +345,40 @@ func (c *CLI) SetupProject() string {
...
+	err = WaitForSelfSAR(1*time.Second, 60*time.Second, c.AdminKubeClient(), kubeauthorizationv1.SelfSubjectAccessReviewSpec{
+		ResourceAttributes: &kubeauthorizationv1.ResourceAttributes{
+			Namespace: newNamespace,
+			Verb:      "create",
+			Group:     "",
+			Resource:  "pods",
+		},
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
```
Comment (deads2k) [inline test/extended/util/client.go:374]: you're missing the standard "wait for these things" from the setupproject.

File: test/extended/util/client.go [hunk 6]
```diff
@@ -124,7 +124,15 @@ func NewCLI(project string) *CLI {
 	cli := NewCLIWithoutNamespace(project)
 	cli.withoutNamespace = false
 	// create our own project
-	g.BeforeEach(func() { _ = cli.SetupProject() })
+	g.BeforeEach(func() {
```
Comment (deads2k) [inline test/extended/util/client.go:127]: once the "wait for these things" step is common between the two impls, then I have relatively few fears about drift given the stability of the projectrequest API.

**Comment by deads2k** [comment]
Currently this PR
1. still relies on openshift APIs not present in microshift
2. does not provide a mechanism to set the user as expected by the test being run
3. does not provide the user with permissions to work in the namespace
4. does not provide interoperability with the functionality it replaces
5. does not provide congruent gating guarantees for all downstream consumers


/hold until the functionality is a drop-in replacement
/assign @ingvagabund

File: test/extended/util/client.go (hunk 3 above)
Comment (pacevedom) [reply test/extended/util/client.go:351]: It works because of https://github.com/openshift/origin/pull/27358/files#diff-e21320d17fdce22cad99f368c20f8aca8b28bac96f890d69e95fda9be0ae7f93R822
Here, `AdminConfig` resolves to the kubeconfig that was used to launch the tests binary, so its the same as the `UserConfig`.
Anyway I can create a function like you say, will follow that approach.

Comment (pacevedom) [reply test/extended/util/client.go:351]: This was covered here: https://github.com/openshift/origin/pull/27358/files#diff-e21320d17fdce22cad99f368c20f8aca8b28bac96f890d69e95fda9be0ae7f93R822

Even though it means that `UserConfig` == `AdminConfig`

File: test/extended/util/client.go (hunk 4 above)
Comment (pacevedom) [reply test/extended/util/client.go:363]: A project creates role bindings for cluster roles `system:image-puller`, `system:image-builder` and `system:deployer`. The first two relate only to OCP APIs, so we can skip them.
The last one I will need it if I change the kubeconfig for the `UserConfig`, as it grants permissions to operate on pods, replication controllers, events and other OCP APIs.

File: test/extended/util/client.go (hunk 3 above)
Comment (deads2k) [reply test/extended/util/client.go:351]: >This was covered here: https://github.com/openshift/origin/pull/27358/files#diff-e21320d17fdce22cad99f368c20f8aca8b28bac96f890d69e95fda9be0ae7f93R822

>Even though it means that UserConfig == AdminConfig

That is functionally incorrect.  The reason we switch users is to prove that the features function properly when run as an unprivileged user.  Given differences in admission and REST handling, using an admin user doesn't test the functionality.

File: test/extended/util/client.go (hunk 4 above)
Comment (deads2k) [reply test/extended/util/client.go:363]: > The last one I will need it if I change the kubeconfig for the UserConfig, as it grants permissions to operate on pods, replication controllers, events and other OCP APIs.

It must switch users or the testing is invalid.  The annotations are important because they change the way that consumers are able to determine ownership.

File: test/extended/util/client.go (hunk 3 above)
Comment (pacevedom) [reply test/extended/util/client.go]: Agree, was not familiar with the project API. Now the service account has the same permissions as a project user.

**Comment by pacevedom** [comment]
Added user certificate handling for authentication.

File: test/extended/util/client.go (hunk 4 above)
Comment (pacevedom) [reply test/extended/util/client.go:371]: Done

File: test/extended/util/client.go (hunk 5 above)
Comment (pacevedom) [reply test/extended/util/client.go:393]: Done

File: test/extended/util/client.go (hunk 3 above)
Comment (pacevedom) [reply test/extended/util/client.go]: Fixed

File: test/extended/util/client.go (hunk 6 above)
Comment (pacevedom) [reply test/extended/util/client.go]: Done

File: test/extended/util/client.go [hunk 7]
```diff
@@ -337,8 +352,155 @@ func (c *CLI) SetupProject() string {
...
+	_, err := c.AdminKubeClient().CoreV1().Namespaces().Create(context.Background(), nsObject, metav1.CreateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+	c.kubeFramework.AddNamespacesToDelete(nsObject)
+
+	framework.Logf("Waiting for ServiceAccount %q to be provisioned...", serviceAccountName)
+	err = WaitForServiceAccount(c.AdminKubeClient().CoreV1().ServiceAccounts(newNamespace), serviceAccountName)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	framework.Logf("Configuring kubeconfig with user %q certificates...", username)
+	c.ChangeUser(username)
```
Comment (deads2k) [inline test/extended/util/client.go:377]: I don't understand how this will work until you modify ChangeUser to tolerate missing the APIs served by the oauth-apiserver and when those APIs are missing use the CSR API to get credentials for the requested user.

Comment (deads2k) [reply test/extended/util/client.go:377]: Found it below

File: test/extended/util/client.go [hunk 8]
```diff
@@ -337,8 +352,160 @@ func (c *CLI) SetupProject() string {
...
+			Name:     "admin",
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "User",
+				Name:      username,
+				Namespace: c.Namespace(),
+			},
+			{
+				Kind:      "ServiceAccount",
```
Comment (deads2k) [inline test/extended/util/client.go:421]: I don't recall granting this power to the service account in the project request case.  ref please?

File: test/extended/util/client.go [hunk 9]
```diff
@@ -337,8 +352,160 @@ func (c *CLI) SetupProject() string {
...
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, err
+	}
+	privateKeyPem := pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	})
+
+	subj := pkix.Name{CommonName: username}
```
Comment (deads2k) [inline test/extended/util/client.go:444]: missing groups?  
1. `system:authenticated`
2. `system:authenticated:oauth`

File: test/extended/util/client.go [hunk 10]
```diff
@@ -337,8 +352,160 @@ func (c *CLI) SetupProject() string {
...
 }
 
+func (c *CLI) setupNamespace() string {
+	requiresTestStart()
+	newNamespace := names.SimpleNameGenerator.GenerateName(fmt.Sprintf("e2e-test-%s-", c.kubeFramework.BaseName))
+	username := fmt.Sprintf("%s-user", newNamespace)
+	serviceAccountName := "default"
+	c.SetNamespace(newNamespace)
+
+	nsObject := &corev1.Namespace{
```
Comment (deads2k) [inline test/extended/util/client.go:362]: missing annotations https://github.com/openshift/openshift-apiserver/blob/master/pkg/project/apiserver/registry/projectrequest/delegated/sample_template.go#L46-L48

File: test/extended/util/client.go (hunk 4 above)
Comment (deads2k) [reply test/extended/util/client.go:371]: still missing annotations.  see https://github.com/openshift/openshift-apiserver/blob/master/pkg/project/apiserver/registry/projectrequest/delegated/sample_template.go#L46-L48

File: test/extended/util/client.go [hunk 11]
```diff
@@ -337,8 +352,160 @@ func (c *CLI) SetupProject() string {
...
+			Group:     "",
+			Resource:  "pods",
+		},
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	err = c.setupNamespacePodSecurity(newNamespace)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	WaitForNamespaceSCCAnnotations(c.KubeClient().CoreV1(), newNamespace)
```
Comment (deads2k) [inline test/extended/util/client.go:397]: update the `WaitForProjectSCCAnnotations` call site to use the `WaitForNamespaceSCCAnnotations` method and remove the `WaitForProjectSCCAnnotations` to prevent accidents

File: test/extended/util/framework.go [hunk 12]
```diff
@@ -1018,32 +1018,60 @@ func CheckBuildCancelled(b *buildv1.Build) bool {
...
+	hasDockerCfg := false
+	for _, s := range sa.ImagePullSecrets {
+		if strings.Contains(s.Name, "-dockercfg-") {
+			hasDockerCfg = true
+		}
+	}
+	return hasDockerCfg
+}
+
+// WaitForServiceAccountWithSecret waits until the named service account gets fully
```
Comment (deads2k) [inline test/extended/util/framework.go:1042]: wrong name

File: test/extended/util/framework.go [hunk 13]
```diff
@@ -1018,32 +1018,60 @@ func CheckBuildCancelled(b *buildv1.Build) bool {
...
+// provisioned. without dockercfg secrets
 func WaitForServiceAccount(c corev1client.ServiceAccountInterface, name string) error {
 	waitFn := func() (bool, error) {
-		sc, err := c.Get(context.Background(), name, metav1.GetOptions{})
+		sa, err := getServiceAccount(c, name)
```
Comment (deads2k) [inline test/extended/util/framework.go:1046]: I see no reason to change this logic

```go
sc, err := c.Get(context.Background(), name, metav1.GetOptions{})
if err != nil {
	// If we can't access the service accounts, let's wait till the controller
	// create it.
	if kapierrs.IsNotFound(err) || kapierrs.IsForbidden(err) {
		e2e.Logf("Waiting for service account %q to be available: %v (will retry) ...", name, err)
		return false, nil
	}
	return false, fmt.Errorf("Failed to get service account %q: %v", name, err)
}
```

is kube-standard (versus nil, nil return) and does what you're trying to do, right?

File: test/extended/util/framework.go [hunk 14]
```diff
@@ -1018,32 +1018,60 @@ func CheckBuildCancelled(b *buildv1.Build) bool {
...
+		return nil, err
+	}
+	return sc, nil
+}
+
+func serviceAccountContainsPullSecret(sa *corev1.ServiceAccount) bool {
+	hasDockerCfg := false
+	for _, s := range sa.ImagePullSecrets {
+		if strings.Contains(s.Name, "-dockercfg-") {
+			hasDockerCfg = true
```
Comment (deads2k) [inline test/extended/util/framework.go:1036]: if you choose to refactor like this, this should be `return true`

File: test/extended/util/framework.go [hunk 15]
```diff
@@ -1018,32 +1018,60 @@ func CheckBuildCancelled(b *buildv1.Build) bool {
...
+		}
+	}
+	return hasDockerCfg
```
Comment (deads2k) [inline test/extended/util/framework.go:1039]: and this should be return false

File: test/extended/util/oc_copy.go [hunk 16]
```diff
@@ -49,7 +49,7 @@ func getUserPartOfNickname(clientCfg *restclient.Config) (string, error) {
 		return "", err
 	}
 	userInfo, err := userClient.Users().Get(context.Background(), "~", metav1.GetOptions{})
-	if kerrors.IsNotFound(err) || kerrors.IsForbidden(err) {
+	if kerrors.IsNotFound(err) || kerrors.IsForbidden(err) || kerrors.IsUnauthorized(err) {
```
Comment (deads2k) [inline test/extended/util/oc_copy.go:52]: why is this necessary?  This is an indication that an invalid token or expired or misconfigured client-certificate was used.  In either case, the error should be exposed.

File: test/extended/util/client.go (hunk 10 above)
Comment (pacevedom) [reply test/extended/util/client.go:363]: Added them.

File: test/extended/util/client.go (hunk 11 above)
Comment (pacevedom) [reply test/extended/util/client.go:403]: Done

File: test/extended/util/framework.go (hunk 13 above)
Comment (pacevedom) [reply test/extended/util/framework.go:1044]: Was just trying to refactor it to avoid repetition, since the same code was in two different functions. Do you think its best to have it self-contained in each of the functions? It is `WaitForServiceAccount` and `WaitForServiceAccountWithSecret`.

File: test/extended/util/client.go (hunk 8 above)
Comment (pacevedom) [reply test/extended/util/client.go]: It slipped through while testing some of the sig-network tests. Removed it.

File: test/extended/util/framework.go (hunk 14 above)
Comment (pacevedom) [reply test/extended/util/framework.go]: Done

File: test/extended/util/framework.go (hunk 15 above)
Comment (pacevedom) [reply test/extended/util/framework.go]: Done

File: test/extended/util/oc_copy.go (hunk 16 above)
Comment (pacevedom) [reply test/extended/util/oc_copy.go]: This was a leftover from the previous version, I removed it.

File: test/extended/util/client.go [hunk 17]
```diff
@@ -331,14 +347,167 @@ func (c *CLI) SetupProject() string {
...
-	WaitForProjectSCCAnnotations(c.ProjectClient().ProjectV1(), newNamespace)
+	WaitForNamespaceSCCAnnotations(c.KubeClient().CoreV1(), newNamespace)
 
 	framework.Logf("Project %q has been fully provisioned.", newNamespace)
 	return newNamespace
 }
 
+func (c *CLI) setupNamespace() string {
+	requiresTestStart()
+	newNamespace := names.SimpleNameGenerator.GenerateName(fmt.Sprintf("e2e-test-%s-", c.kubeFramework.BaseName))
```
Comment (stlaz) [inline test/extended/util/client.go:358]: why don't you use the kube framework method for namespace creation? It's likely to do most work for you.

File: test/extended/util/client.go (hunk 10 above)
Comment (stlaz) [reply test/extended/util/client.go:363]: @deads2k if these are not added in Microshift, why do we add them? The tests that require them should rightfully fail in our case, no?

File: test/extended/util/client.go (hunk 9 above)
Comment (pacevedom) [reply test/extended/util/client.go:445]: Got this wrong, let me rework it.

Comment (pacevedom) [reply test/extended/util/client.go]: Done

File: test/extended/util/client.go [hunk 18]
```diff
@@ -251,6 +257,15 @@ func (c CLI) WithToken(token string) *CLI {
 // All resources will be then created within this project.
 // Returns the name of the new project.
 func (c *CLI) SetupProject() string {
+	exist, err := DoesApiResourceExist(c, "projects")
```
[the remark this answers is not in the document]
Comment (deads2k) [reply test/extended/util/client.go:261]: > Wouldn't it make more sense if this method accepted the resource and its group?

We did not instrument with individual APIs.

File: test/extended/util/client.go (hunk 10 above)
Comment (deads2k) [reply test/extended/util/client.go:363]: > @deads2k if these are not added in Microshift, why do we add them? The tests that require them should rightfully fail in our case, no?

Possibly, though I think we can say, "make an equivalent namespace"?  I could also get behind trying to tighten once we break the back of the problem.

File: test/extended/util/framework.go [hunk 19]
```diff
@@ -1018,87 +1017,88 @@ func CheckBuildCancelled(b *buildv1.Build) bool {
 	return b.Status.Phase == buildv1.BuildPhaseCancelled
 }
 
+func getServiceAccount(c corev1client.ServiceAccountInterface, name string) (*corev1.ServiceAccount, error) {
+	sc, err := c.Get(context.Background(), name, metav1.GetOptions{})
+	if err != nil {
+		if kapierrs.IsNotFound(err) || kapierrs.IsForbidden(err) {
+			return nil, nil
```
Comment (deads2k) [inline test/extended/util/framework.go:1024]: This is a blocker.  We don't do this when working with kube APIs because it defies expectations about client handling that have existed for the past eight years.

File: test/extended/util/framework.go (hunk 13 above)
Comment (deads2k) [reply test/extended/util/framework.go:1044]: still a problem here.

File: test/extended/util/framework.go (hunk 19 above)
Comment (pacevedom) [reply test/extended/util/framework.go]: Left it as it was before.

**Comment by deads2k** [comment]
no conflict with the PR that merged 12h ago.
